jdrfandroidbleparser/src/main/java/org/ehealthinnovation/jdrfandroidbleparser/bgm/utility/BluetoothDateTime.kt

```diff
@@ -0,0 +1,34 @@
+package org.ehealthinnovation.jdrfandroidbleparser.bgm.utility
+
+import java.util.*
+
+class BluetoothDateTime {
+    companion object {
+        /**
+         * Convert bluetooth date time fields into Java Date format. It takes care of the offset addition.
+         * @param y Year in bluetooth definition
+         * @param m Month in bluetooth definition
+         * @param d Day in bluetooth definition
+         * @param hr Hour in bluetooth definition
+         * @param min Minute in bluetooth definition
+         * @param sec Second in blutooth definition
+         * @param offset the offset
+         * @return the date of the measurement
+         * @see https://www.bluetooth.com/specifications/gatt/viewer?attributeXmlFile=org.bluetooth.characteristic.date_time.xml
+         */
+        fun getDate(y:Int?, m:Int?, d:Int?, hr:Int?, min:Int?, sec:Int?, offset:Int?): Date? {
+            if(y==null || m==null || d==null || hr==null || min==null || sec==null)
+                return null
+
+            val calendar = Calendar.getInstance()
+            calendar.set(Calendar.MILLISECOND, 0)
+            //0 means the field is unknown in bluetooth
+            if(y==0 || m==0 || d==0 )
+                return null
+
+            calendar.set(y, m-1, d, hr, min, sec)
+            offset?.let { calendar.add(Calendar.MINUTE, it)}
+            return calendar.time
+        }
+    }
+}
```

Comment:
So, right now...this is a giant get method that doesn't actually do any getting, it calculates the date based on the passed in values...which could be null? or 0?

There is definitely a better cleaner way to go about this. Maybe a builder pattern? or overloading the constructor to add evaluations of the passed in params. There are lots of ways to go about this.

Reply:
Make it a data class to start

Reply:
This intends to be a conversion method, to convert Bluetooth date values into Java Date class. It was originally within the GlucoseMeasurementCharacteristic Class, but I think multiple characteristic can use it later, so I singled it out as a standalone static helper class.

Reply:
If it's not something related to parsing to and from the data class, then it shouldn't be here. If you want a util class, we can talk about designing that, but we're not going to be injecting UI helper classes into the lib. This lib doesn't care about UI.

Reply:
I think there is some misunderstanding here. DisplayTime is not referring to the UI aspect of display. It is the time value that can be used in displayed to the user (a user friendly time) as oppose to machine time which contains date field and time offset. I changed the name to userFacingTime in the revise version.

Reply:
Display time provides a user based component in a data class. It needs to be removed. When we do UI, we can do another utility or class, or even add it back into here...if we deem it's necessary at the time. But right now, this class doesn't care about UI, or users, or really anything other than parsing. Separation of concerns.

Reply:
The get time method was redundant and is removed. At the same time, the convert time utility is kept to provide a further step in parsing. User facing time is noted in the specification, therefore we are providing tools to pull out what the specification mentions into usable Java data type with minimum effort. We know that the raw year, month, day etc can not be used directly by the consumers of this library, and they are bound to look into the spec and write a conversion layer if this method was not present. The user should not bare the burden of interpreting the spec and the risk of mis-implementing the spec.